cloud/aws/actuators/cluster/context.go

```diff
@@ -0,0 +1,72 @@
+// Copyright © 2018 The Kubernetes Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	providerconfigv1 "sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	client "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/typed/cluster/v1alpha1"
+)
+
+// Context defines the context services accept and work within.
+type Context interface {
+	Cluster() *clusterv1.Cluster
+	Client() client.ClusterInterface
+	Config() *providerconfigv1.AWSClusterProviderConfig
+	Status() *providerconfigv1.AWSClusterProviderStatus
+}
+
+// contextFromCluster returns a new cluster context.
+func (a *Actuator) contextFromCluster(cluster *clusterv1.Cluster) (Context, error) {
```

Comment:
This should return a concrete type, not an interface. This is an antipattern that kubernetes follows and we should not follow that pattern.

Please see this paragraph https://peter.bourgon.org/go-for-industrial-programming/#how-much-interface-do-i-need

I will say defining interfaces is fine when other people have to implement an interface you define. For example, the sort package does this.